libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -67,28 +82,40 @@ public new BotFrameworkAdapter Use(IMiddleware middleware)
                 {
                     // The Activity Schema doesn't have a delay type build in, so it's simulated
                     // here in the Bot. This matches the behavior in the Node connector. 
-                    int delayMs = (int)((Activity)activity).Value;
+                    int delayMs = (int)activity.Value;
                     await Task.Delay(delayMs).ConfigureAwait(false);
                 }
                 else
                 {
-                    MicrosoftAppCredentials appCredentials = await GetAppCredentials((context as BotFrameworkBotContext).BotAppId);
+                    BotFrameworkBotContext bfContext = context as BotFrameworkBotContext;
+                    if (bfContext == null)
+                        throw new InvalidOperationException($"BotFramework Context is required. Incorrect context type: {context.GetType().Name}");
+
+                    MicrosoftAppCredentials appCredentials = await GetAppCredentials(bfContext.BotAppId);
                     var connectorClient = new ConnectorClient(new Uri(activity.ServiceUrl), appCredentials);
-                    await connectorClient.Conversations.SendToConversationAsync((Activity)activity).ConfigureAwait(false);
+                    await connectorClient.Conversations.SendToConversationAsync(activity).ConfigureAwait(false);
                 }
             }
         }
 
         public override async Task<ResourceResponse> UpdateActivity(IBotContext context, Activity activity)
         {
-            MicrosoftAppCredentials appCredentials = await GetAppCredentials((context as BotFrameworkBotContext).BotAppId);
+            BotFrameworkBotContext bfContext = context as BotFrameworkBotContext;
+            if (bfContext == null)
+                throw new InvalidOperationException($"BotFramework Context is required. Incorrect context type: {context.GetType().Name}");
+
+            MicrosoftAppCredentials appCredentials = await GetAppCredentials(bfContext.BotAppId);
             var connectorClient = new ConnectorClient(new Uri(activity.ServiceUrl), appCredentials);
-            return await connectorClient.Conversations.UpdateActivityAsync((Activity)activity);
+            return await connectorClient.Conversations.UpdateActivityAsync(activity);
         }
 
         public override async Task DeleteActivity(IBotContext context, ConversationReference reference)
         {
-            MicrosoftAppCredentials appCredentials = await GetAppCredentials((context as BotFrameworkBotContext).BotAppId);
+            BotFrameworkBotContext bfContext = context as BotFrameworkBotContext;
+            if (bfContext == null)
```

Comment:
Worth moving this to a separate method? This same code is being repeated at multiple places.

Reply:
Yea, I'm actually hoping to change that today. 

Node already has it, and I wanted to get the "Fire and Forget" version(s) working first as the initial M2 changeset was so large I figured I would just do it after the main codebase was updated.

Reply:
👍